codalab/server/bundle_manager.py

```diff
@@ -714,7 +739,27 @@ def _get_staged_bundles_to_run(self):
                     bundle,
                     {'state': State.FAILED, 'metadata': {'failure_message': failure_message}},
                 )
+            elif bundle.metadata.request_queue and len(matched_workers) == 0:
+                # For those bundles that were requested to run on a private worker, if the private worker does not exist
+                # temporarily, we filter out those bundles so that they won't be dispatched to run on workers.
+                if bundle.uuid not in self._bundles_wo_matched_workers:
+                    # TODO: update bundles with batch operation: implement function batch_update_bundles()
+                    self._model.update_bundle(
+                        bundle,
+                        {
+                            'metadata': {
+                                'staged_status': 'Bundle is requested to run on the private worker {} which has '
+                                'not been connected to the CodaLab server yet.'.format(
+                                    request_queue
+                                )
+                            }
+                        },
+                    )
+                    self._bundles_wo_matched_workers.add(bundle.uuid)
```

Comment:
This state is only for efficiency, right?  Did we check that it makes a difference in practice?  Again, try to get rid of state as much as possible.

Reply:
I agree that we shouldn't make special state as much as possible. I think what we are doing here is based on a point brought up earlier which is to avoid updating the database with the exact same information. In reality, the number of bundles that don't have private workers matched could vary from time to time. And this is something we can't control. By timing the `update_bundle` function, one query takes `0.0023462772369384766 seconds` when the database is in the same instance as rest-server/bundle-manager/worker. In production, this number will be larger for sure, as the database is hosted in a separate machine. Technically this will improve the performance for sure. But in terms of how much it will improve the performance overall will fully depend on how many bad bundles submitted to the system.